src/sql/src/kafka_util.rs

```diff
@@ -246,6 +250,132 @@ pub async fn test_config(
     Ok(())
 }
 
+/// Returns start offsets for the partitions of `topic` and the provided
+/// `time_offset` option.
+///
+/// For each partition, the returned offset is the earliest offset whose
+/// timestamp is greater than or equal to the given timestamp for the
+/// partition. If no such message exists (or the Kafka broker is before
+/// 0.10.0), the current end offset is returned for the partition.
+///
+/// The provided `time_offset` option must be a non-zero number:
+/// * Non-Negative numbers will used as is (e.g. `1622659034343`)
+/// * Negative numbers will be translated to a timestamp in millis
+///   before now (e.g. `-10` means 10 millis ago)
+///
+/// If `time_offset` has not been configured, an empty Option is returned.
+pub async fn lookup_start_offsets(
+    broker: &str,
+    topic: &str,
+    client_options: &BTreeMap<String, String>,
+    with_options: &BTreeMap<String, Value>,
+) -> Result<Option<Vec<i64>>, anyhow::Error> {
+    let time_offset = with_options.get("time_offset");
+    if time_offset.is_none() {
+        return Ok(None);
+    } else if with_options.contains_key("start_offset") {
+        bail!("`start_offset` and `time_offset` cannot be set at the same time.")
+    }
+
+    // Validate and resolve `time_offset`.
+    let time_offset = match time_offset.unwrap() {
+        Value::Number(s) => match s.parse::<i64>() {
+            // Timestamp in millis *before* now (e.g. -10 means 10 millis ago)
+            Ok(ts) if ts < 0 => {
+                let now: i64 = SystemTime::now()
+                    .duration_since(SystemTime::UNIX_EPOCH)?
+                    .as_millis()
+                    .try_into()?;
+                let ts = now - ts.abs();
+                if ts <= 0 {
+                    bail!("Relative `time_offset` must be smaller than current system timestamp")
+                }
+                ts
+            }
+            // Timestamp in millis (e.g. 1622659034343)
+            Ok(ts) => ts,
+            _ => bail!("`time_offset` must be a number"),
+        },
+        _ => bail!("`time_offset` must be a number"),
+    };
+
+    // Lookup offsets
+    let mut config = rdkafka::ClientConfig::new();
+    config.set("bootstrap.servers", broker);
+    for (k, v) in client_options {
+        config.set(k, v);
+    }
+
+    let consumer: BaseConsumer<KafkaErrCheckContext> =
+        match config.create_with_context(KafkaErrCheckContext::default()) {
+            Ok(consumer) => consumer,
+            Err(e) => {
+                match e {
+                    rdkafka::error::KafkaError::ClientCreation(s) => {
+                        // Pass existing error back up.
+                        bail!(rdkafka::error::KafkaError::ClientCreation(s))
+                    }
+                    _ => bail!(e),
+                }
+            }
+        };
+
+    task::spawn_blocking({
```

Comment:
I was unsure whether we prefer multiple smaller scoped calls to `spawn_blocking` for the individual requests or just have one.

Reply:
I'm also unsure! The only other call to `spawn_blocking` that I see is in [`test_config`](https://github.com/MaterializeInc/materialize/blob/main/src/sql/src/kafka_util.rs#L215), which doesn't answer our question. @petrosagg or @benesch would probably have the most useful advice here.

Reply:
My model is that every call to `spawn_blocking` has approximately as much overhead as creating a new thread. You can think of `spawn_blocking` as equivalent to:

```rust
let (tx, rx) = std::sync::mpsc::channel();
thread::spawn({
    let result = /* do stuff */;
    tx.send(result);
});
rx.recv();
```

So that's an argument for the general rule being "use fewer calls to `spawn_blocking`", but also feel free to make your own determination based on the situation.

Reply:
I didn't know the answer to this so I looked at the implementation and ran a small experiment. I think it's ok to have many **sequencial** calls to `spawn_blocking`. Tokio spawns blocking threads on demand but crucially, it keeps them around for some time (10 seconds by default) and reuses them for subsequent blocking needs.

This test program spawns just two extra threads:

```rust
use std::{thread, time::Duration};

#[tokio::main]
async fn main() {
    println!("runtime started");
    for _ in 0..100 {
        let _ = tokio::task::spawn_blocking(|| {
            thread::sleep(Duration::from_millis(10));
        }).await;
    }
    println!("blocking done");
    thread::sleep(Duration::from_millis(10_010));
    println!("workers should be gone now");
}
```

`strace` output:

```rust
strace -f -e trace=clone ./target/debug/tokio-block-test
clone(child_stack=0x7f47c35fedb0, [...]) = 338157
strace: Process 338157 attached
[pid 338156] clone(child_stack=0x7f47c33fddb0, [..]) = 338158
strace: Process 338158 attached
[pid 338156] clone(child_stack=0x7f47c31f9db0, [..]) = 338159
strace: Process 338159 attached
[pid 338156] clone(child_stack=0x7f47c2ff5db0, [..]) = 338160
strace: Process 338160 attached
runtime started
[pid 338156] clone(child_stack=0x7f47c2df4db0, [..]) = 338161
strace: Process 338161 attached
[pid 338156] clone(child_stack=0x7f47c2beadb0, [..]) = 338162
strace: Process 338162 attached
blocking done
[pid 338161] +++ exited with 0 +++
[pid 338162] +++ exited with 0 +++
workers should be gone now
[pid 338157] +++ exited with 0 +++
[pid 338158] +++ exited with 0 +++
[pid 338159] +++ exited with 0 +++
[pid 338160] +++ exited with 0 +++
+++ exited with 0 +++
```

Reply:
Thank you very much for your inputs!  I'll stick to the single `spawn_blocking` call here since the blocking API calls are very close to each other and not much happens in between calls. Still, good to know that the threads would likely be re-used and the mental model is very helpful, too. 🚀